File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/quaternion.py
"""Generated mapper for `geometry_msgs/msg/Quaternion`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.geometry_msgs.msg.v1 import Quaternion as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def quaternion_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.x = msg.x
    bus.y = msg.y
    bus.z = msg.z
    bus.w = msg.w
    return bus


def quaternion_to_ros(bus):
    from geometry_msgs.msg import Quaternion as RosMsg

    out = RosMsg()
    out.x = bus.x
    out.y = bus.y
    out.z = bus.z
    out.w = bus.w
    return out


class GeometryMsgsQuaternionMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from geometry_msgs.msg import Quaternion as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return quaternion_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return quaternion_to_ros(bus)

File: src/ros2_bridge/mappers/geometry_msgs/quaternion.rs
//! Typed mapper for `geometry_msgs/msg/Quaternion`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn quaternion_to_bus(
    msg: ros_env::geometry_msgs::msg::Quaternion,
) -> crate::geometry_msgs::msg::v1::Quaternion {
    crate::geometry_msgs::msg::v1::Quaternion {
        x: msg.x,
        y: msg.y,
        z: msg.z,
        w: msg.w,
    }
}

pub(crate) fn quaternion_to_ros(
    bus: crate::geometry_msgs::msg::v1::Quaternion,
) -> ros_env::geometry_msgs::msg::Quaternion {
    ros_env::geometry_msgs::msg::Quaternion {
        x: bus.x,
        y: bus.y,
        z: bus.z,
        w: bus.w,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsQuaternionMapper;

impl TypedTopicMapper for GeometryMsgsQuaternionMapper {
    type Ros = ros_env::geometry_msgs::msg::Quaternion;
    type Bus = crate::geometry_msgs::msg::v1::Quaternion;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(quaternion_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(quaternion_to_ros(msg))
    }
}
